### PR TITLE
[Windows] Serialize the processing of true key events and simulated key events generated by Keyman

### PR DESCRIPTION
**Pull request by mcdurdin**
Serializes all input events onto the current focused window thread by repeating any externally generated events. This is required to ensure that we can get the correct modifier state when we need to push a modifier release/set around keystroke output (mostly for the purposes of the backspace event).

We need to release Alt and Ctrl modifiers (and we do Shift as well for completeness) when emitting Backspace to modify context in order to avoid triggering Alt+Backspace -> Undo or Ctrl+Backspace -> 0x7F / Word delete left instead of a Char delete left event.

However it is possible that a modifier release event generated by the user is already in the queue at the time we send the input. Since we have no way to detect or prevent this happening, what we do instead is re-inject each keystroke into the queue in the focused input thread, which means we can guarantee order of events and sort out the modifier state as required.  This looks really messy in the message event queue but turns out to be robust in practice.

You can disable this flag with flag_ShouldSerializeInput and flag_IsDevEnvProcess handles a nuance with devenv.exe.

Note: I really dislike having to handle the nuance because who is to say this is the only one? But it's the only way I can see to work around what really is an edge case. Nothing is ever completely straightforward!

I would like to see this change getting thorough dogfood testing because it is a significant update to the keystroke processing and may have other unpredictable side effects.

### CODE REVIEW

File: windows/src/engine/keyman32/keybd_shift.cpp [hunk 1]
```diff
@@ -105,33 +105,16 @@ void do_keybd_event(LPINPUT pInputs, int *n, BYTE vk, BYTE scan, DWORD flags, UL
   in the current Keyman keyboard, because without the dummy prefix key, faking the release of the 
   Alt key will activate the mainmenu of the current application.
 */
-void keybd_sendprefix(LPINPUT pInputs, int *n, BOOL *FPrefix)
+void keybd_sendprefix(LPINPUT pInputs, int *n)
```
Comment (mcdurdin) [inline windows/src/engine/keyman32/keybd_shift.cpp]: I have refactored a lot of the code here -- it looks a bit complicated but actually simplifies the use considerably. See `keybd_shift()` for the entry point to all these functions.

File: windows/src/engine/keyman32/kmhook_getmessage.cpp [hunk 2]
```diff
@@ -141,32 +142,79 @@ LRESULT _kmnGetMessageProc(int nCode, WPARAM wParam, LPARAM lParam)
 		if(!InitialiseProcess(mp->hwnd)) return CallNextHookEx(Globals::get_hhookGetMessage(), nCode, wParam, lParam);
   }
 
-	if(mp->message == WM_CHAR || mp->message == WM_SYSCHAR || mp->message == WM_KEYDOWN || mp->message == WM_SYSKEYDOWN || mp->message == WM_KEYUP) 
+	if(mp->message == WM_CHAR || mp->message == WM_SYSCHAR || mp->message == WM_KEYDOWN || mp->message == WM_SYSKEYDOWN || mp->message == WM_KEYUP || mp->message == WM_SYSKEYUP)
```
Comment (darcywong00) [inline windows/src/engine/keyman32/kmhook_getmessage.cpp]: tab indent?

File: windows/src/engine/keyman32/keybd_shift.cpp [hunk 3]
```diff
@@ -145,26 +128,48 @@ void GetAsyncKeyboardShiftState(BYTE *kbd) {
               FReset   are we clearing or resetting the modifier state?
```
Comment (darcywong00) [inline windows/src/engine/keyman32/keybd_shift.cpp]: `FReset` is now `isReset`

File: windows/src/engine/keyman32/keybd_shift.cpp [hunk 4]
```diff
@@ -145,26 +128,48 @@ void GetAsyncKeyboardShiftState(BYTE *kbd) {
...
-
-  There is a potential for a race here because the user may release a modifier
-  key between our test for the key state and when we SendInput, in which case the key will be 
-  incorrectly re-pressed in the reset phase. This window of time is very small; we can improve the
-  odds by using GetAsyncKeyboardShiftState (which reads the key state at the current instant, as 
-  opposed to at the start of the current key event).
 */
-void keybd_shift(LPINPUT pInputs, int *n, BOOL FReset, LPBYTE kbd) {
+
+void keybd_shift_release(LPINPUT pInputs, int *n, LPBYTE kbd) {
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply windows/src/engine/keyman32/keybd_shift.cpp]: I think I'd prefer to improve names over time -- `kbd_shift_release` and `kbd_shift_reset` are more namespaced -- the whole kit and caboodle should really be a class but that's a refactor too far for today.

**Pull request merged**